core/src/main/java/org/elasticsearch/index/query/support/InnerHitBuilder.java

```diff
@@ -196,19 +227,93 @@ public void writeTo(StreamOutput out) throws IOException {
         }
         out.writeOptionalWriteable(highlightBuilder);
         out.writeNamedWriteable(query);
-        out.writeOptionalWriteable(innerHitsBuilder);
+        boolean hasChildInnerHits = childInnerHits != null;
+        out.writeBoolean(hasChildInnerHits);
+        if (hasChildInnerHits) {
+            out.writeVInt(childInnerHits.size());
+            for (Map.Entry<String, InnerHitBuilder> entry : childInnerHits.entrySet()) {
+                out.writeString(entry.getKey());
+                entry.getValue().writeTo(out);
+            }
+        }
     }
 
     public InnerHitBuilder setParentChildType(String parentChildType) {
         this.parentChildType = parentChildType;
         return this;
     }
 
+    public InnerHitBuilder setParentChildType(String parentChildType, QueryBuilder<?> qb) {
+        this.parentChildType = parentChildType;
+        this.query = qb;
+        if (name == null) {
+            name = parentChildType;
+        }
+        inlineLeafInnerHits(qb);
+        return this;
+    }
+
     public InnerHitBuilder setNestedPath(String nestedPath) {
         this.nestedPath = nestedPath;
         return this;
     }
 
+    public InnerHitBuilder setNestedPath(String nestedPath, QueryBuilder<?> qb) {
+        this.nestedPath = nestedPath;
+        this.query = qb;
+        if (name == null) {
+            name = nestedPath;
+        }
+        inlineLeafInnerHits(qb);
+        return this;
+    }
+
+    // Inlines the inner hits of leaf queries with this inner hit,
+    // this is for hierarchical inner hits
+    void inlineLeafInnerHits(QueryBuilder<?> qb) {
+        if (qb instanceof HasChildQueryBuilder) {
+            HasChildQueryBuilder hcqb = (HasChildQueryBuilder) qb;
+            InnerHitBuilder ihb = hcqb.innerHit();
+            if (ihb != null) {
+                addChildInnerHit(ihb);
+                hcqb.clearInnerHit();
+            }
+        } else if (qb instanceof HasParentQueryBuilder) {
+            HasParentQueryBuilder hpqb = (HasParentQueryBuilder) qb;
+            InnerHitBuilder ihb = hpqb.innerHit();
+            if (ihb != null) {
+                addChildInnerHit(ihb);
+                hpqb.clearInnerHit();
+            }
+        } else if (qb instanceof NestedQueryBuilder) {
+            NestedQueryBuilder nqb = (NestedQueryBuilder) qb;
+            InnerHitBuilder ihb = nqb.innerHit();
+            if (ihb != null) {
+                addChildInnerHit(ihb);
+                nqb.clearInnerHit();
+            }
+        } else if (qb instanceof ConstantScoreQueryBuilder) {
```

Comment:
All these instanceof calls look fragile to me. I'm wondering that query builders could have an API to expose their sum query builders instead?